### ISSUE TITLE
Client hangs forever disconnecting from an unstable connection

### ISSUE THREAD

**Issue by ctron** [eclipse-paho/paho.mqtt.java#330]
Please fill out the form below before submitting, thank you!

- [X] Bug exists Release Version 1.1.0 ( Master Branch)
- [X] Bug exists in Snapshot Version 1.1.1-SNAPSHOT (Develop Branch)

I do have a broker which accepts connections but then rejects the logon. Using Paho with the async connection in auto re-connect mode I am unable to properly shut down the client. Most of the times (not always) the client gets stuck in:

```
Stack trace: 
java.lang.Object.wait(Native Method)
java.lang.Object.wait(Object.java:502)
org.eclipse.paho.client.mqttv3.internal.Token.waitForResponse(Token.java:141)
org.eclipse.paho.client.mqttv3.internal.Token.waitForCompletion(Token.java:108)
org.eclipse.paho.client.mqttv3.MqttToken.waitForCompletion(MqttToken.java:63)
org.eclipse.kapua.kura.simulator.MqttSimulatorTransport.close(MqttSimulatorTransport.java:161)
org.eclipse.kapua.its.base.BasicTest.test1(BasicTest.java:101)
```

The call that I make in order to close is:

```java
public void close() throws MqttException {
	try {
		this.client.disconnect(0).waitForCompletion();
	} finally {
		this.client.close();
	}
}
```

**Comment by ctron** [comment]
There seems to be a simple reproducer:

```java
import org.eclipse.paho.client.mqttv3.MqttAsyncClient;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;

public class Application {
	public static void main(String[] args) throws Exception {
		MqttAsyncClient client = new MqttAsyncClient("tcp://iot.eclipse.org:1882", "client-1");

		MqttConnectOptions options = new MqttConnectOptions();
		options.setAutomaticReconnect(true);
		options.setUserName("foo");
		options.setPassword("bar".toCharArray());
		client.connect(options);

		Thread.sleep(1_000);

		try {
			client.disconnect(0).waitForCompletion();
		} finally {
			client.close();
		}
	}
}
```

Please note, the port number 1882 is on purpose.

**Comment by ctron** [comment]
The re-producer has to be changed to:

```java
import org.eclipse.paho.client.mqttv3.MqttAsyncClient;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;

public class Application {
	public static void main(String[] args) throws Exception {
		MqttAsyncClient client = new MqttAsyncClient("tcp://iot.eclipse.org:1882", "client-1");

		MqttConnectOptions options = new MqttConnectOptions();
		options.setAutomaticReconnect(true);
		options.setUserName("foo");
		options.setPassword("bar".toCharArray());
		options.setConnectionTimeout(2);
		client.connect(options);

		Thread.sleep(1_000);

		try {
			client.disconnect(0).waitForCompletion();
		} finally {
			client.close();
		}
	}
}
```

Otherwise you are running in a different issue where the socket connect call hangs the disconnect process.

**Comment by carlos-lacerda** [comment]
Hi @ctron , I'm using the client mqtt.java version 1.0.2 and facing this same error you have found. 
I'm looking for a workaround before your PR with fix be accepted. In you case, you builded your own version with fix to solve the problem or did other approach? I'm thinking about to use your fix and build a temporary version or maybe start an anonymous thread that tries close the connection or do the whole thing connect/send/close that not solves the problem but at least not hang my app. Thank you for any tip.

**Comment by ctron** [comment]
@carlos-lacerda Yes, I am using a locally built version.

I didn't have much time right now to look into this. But I probably will do this in the nearer future.

**Comment by larrywalker** [comment]
Any ideas when this might find it’s way into a release?

I was running v1.1.0 and noticed a number of leaked threads. I updated to v1.1.1 and notice the same issue.

"MQTT Disc: clientid1503079649368000000" #379 daemon prio=5 os_prio=0 tid=0x00007fa25c0f5800 nid=0xe97 in Object.wait() [0x00007fa238f9f000]
   java.lang.Thread.State: WAITING (on object monitor)
        at java.lang.Object.wait(Native Method)
        at java.lang.Object.wait(Object.java:502)
        at org.eclipse.paho.client.mqttv3.internal.Token.waitUntilSent(Token.java:248)
        - locked <0x00000006c284af20> (a java.lang.Object)
        at org.eclipse.paho.client.mqttv3.internal.ClientComms$DisconnectBG.run(ClientComms.java:727)
        at java.lang.Thread.run(Thread.java:748)

**Comment by tatemz** [comment]
Any update on this issue? I am also seeing this issue in v1.2.0

**Comment by larrywalker** [comment]
Starting to give up hope on an official patch/release, so I pulled down the tagged v1.1.1 sources the other day and manually applied the proposed patch for this issue.  The resulting jar was worse as it leaks the normal communication threads worse than any of the previous versions.  

I didn't notice any "MQTT Disc" threads getting leaked, but the version that I build was not trustworthy.

**Comment by ctron** [comment]
@larrywalker Well it worked in my specific scenario. But I never got the unit tests to work properly, so yes, there may be a bunch of untested scenarios which fail. After no responses from Paho for quite I while I switched to Fuse MQTT instead.

I think the best is to close out the PR and hope for some future update.

**Comment by jpwsutton** [comment]
Hey folks, sorry for not getting round to this. Currently I'm the only main developer on this project as well as a few others and I've been quite busy so haven't been able to take a deeper look at this issue. I'd say to keep the PR open for now as it's a genuine issue that needs to be fixed. I'll do my very best to do some investigation on this problem soon, but would definitely welcome others to help out as well if they are able.

**Comment by tatemz** [comment]
I resolved the `waitForResponse` error that you posted above. See my comment on #207 

https://github.com/eclipse/paho.mqtt.java/issues/207#issuecomment-338246879

However, 1.2.0 still shows a memory leak that I have not traced yet. I will attempt this patch manually and test.

**Comment by icraggs** [comment]
I'm looking at this now.   Basically, the problem is that the disconnect is started while the connect process hasn't yet finished.  Potential solutions:

1) Return "connect in progress" from the disconnect call.  Not particularly helpful, but it is the same approach as is taken when calling disconnect and the client is disconnected already, or in the process of disconnecting.  

2) "Queue" up the disconnect request until the current connect request has finished or can be interrupted (if possible).  In the simplest approach, in the scenario in this test, this would mean that the disconnect request wouldn't start until the connect had timed out, so it could be a while.

Actually, approach 1) would indicate to the developer that perhaps the application should best wait until the connect operation is complete, one way or another, before attempting a disconnect if necessary.

**Comment by icraggs** [comment]
In the fix I've added, I haven't taken either of those two approaches (yet).  There was a missing signal when the sender process was not running.  I still might add the connect in progress return code at some point.

**Comment by icraggs** [comment]
Fixed released in 1.2.1
